src/Models/HydrostaticFreeSurfaceModels/mg_implicit_free_surface_solver.jl

```diff
@@ -0,0 +1,171 @@
+using Oceananigans.Solvers
+using Oceananigans.Operators
+using Oceananigans.ImmersedBoundaries: ImmersedBoundaryGrid, GridFittedBottom
+using Oceananigans.Architectures
+using Oceananigans.Grids: with_halo, isrectilinear
+using Oceananigans.Fields: Field, ZReducedField
+using Oceananigans.Architectures: device, unsafe_free!
+using Oceananigans.Models.HydrostaticFreeSurfaceModels: Az_∇h²ᶜᶜᶜ
+using Oceananigans.Solvers: fill_matrix_elements!
+using Oceananigans.Solvers: constructors, arch_sparse_matrix, ensure_diagonal_elements_are_present!, update_diag!
+
+import Oceananigans.Solvers: solve!, precondition!
+import Oceananigans.Architectures: architecture
+
+"""
+    struct MGImplicitFreeSurfaceSolver{V, S, R}
+
+The multigrid implicit free-surface solver.
+
+$(TYPEDFIELDS)
+"""
+mutable struct MGImplicitFreeSurfaceSolver{S, V, F, R, C, D}
+    "The multigrid solver"
+    multigrid_solver :: S
+    "The vertically-integrated lateral areas"
+    vertically_integrated_lateral_areas :: V
+    "The previous time step"
+    previous_Δt :: F
+    "The right hand side of the free surface evolution equation"
+    right_hand_side :: R
+    "The matrix constructors of the linear operator without the Az / (g * Δt^2) term"
+    matrix_constructors :: C
+    "The Az / g term"
```

Comment:
```suggestion
    "The `Az / g` term"
```

is it `Az / g` or `Az / (g Δt²)`?

Reply:
Just `Az / g`, just storing this value for all `i` and `j` so it's easier to add on `-Az / (g Δt²)` later, to replicate the [`diagonal ` field ](https://github.com/CliMA/Oceananigans.jl/blob/80f328b9d94bb4f13af8347f06356b93127638cc/src/Solvers/heptadiagonal_iterative_solver.jl#L14) in the heptadiagonal solver which in the implicit free surface case is initialised [here](https://github.com/CliMA/Oceananigans.jl/blob/80f328b9d94bb4f13af8347f06356b93127638cc/src/Models/HydrostaticFreeSurfaceModels/matrix_implicit_free_surface_solver.jl#L136).

Reply:
Gotcha!